sigstore/dsse.py

```diff
@@ -0,0 +1,184 @@
+        from constituent pieces.
+        """
+        self._contents = contents
+        try:
+            self._statement = _Statement.model_validate_json(contents)
+        except ValidationError:
+            raise ValueError("malformed in-toto statement")
+
+    def _pae(self) -> bytes:
+        """
+        Construct the PAE encoding for this statement.
+        """
+
+        # See:
+        # https://github.com/secure-systems-lab/dsse/blob/v1.0.0/envelope.md
+        # https://github.com/in-toto/attestation/blob/v1.0/spec/v1.0/envelope.md
+        pae = f"DSSEv1 {len(Statement._ENVELOPE_TYPE)} {Statement._ENVELOPE_TYPE} ".encode()
+        pae += b" ".join([str(len(self._contents)).encode(), self._contents])
+        return pae
+
+    def sign(self, key: ec.EllipticCurvePrivateKey) -> Envelope:
+        """
+        Sign the statement, returning a DSSE envelope containing the statement's
+        signature.
+        """
+
+        pae = self._pae()
+        logger.debug(f"DSSE PAE: {pae!r}")
+
+        signature = key.sign(pae, ec.ECDSA(hashes.SHA256()))
+        return Envelope(
+            payload=self._contents,
+            payload_type=Statement._ENVELOPE_TYPE,
+            signatures=[Signature(sig=signature, keyid=None)],
+        )
+
+
+class _StatementBuilder:
+    """
+    A builder-style API for constructing in-toto Statements.
+    """
+
+    def __init__(
+        self,
+        subjects: list[_Subject] | None = None,
+        predicate_type: str | None = None,
+        predicate: dict[str, Any] | None = None,
+    ):
+        """
+        Create a new `_StatementBuilder`.
+        """
+        self._subjects = subjects or []
+        self._predicate_type = predicate_type
+        self._predicate = predicate
+
+    def subjects(self, subjects: list[_Subject]) -> _StatementBuilder:
+        """
+        Configure the subjects for this builder.
+        """
+        self._subjects = subjects
+        return self
+
+    def predicate_type(self, predicate_type: str) -> _StatementBuilder:
+        """
+        Configure the predicate type for this builder.
+        """
+        self._predicate_type = predicate_type
+        return self
+
+    def predicate(self, predicate: dict[str, Any]) -> _StatementBuilder:
+        """
+        Configure the predicate for this builder.
+        """
+        self._predicate = predicate
+        return self
+
+    def build(self) -> Statement:
+        """
+        Build a `Statement` from the builder's state.
+        """
+        try:
+            stmt = _Statement(
+                type_="https://in-toto.io/Statement/v1",
+                subjects=self._subjects,
+                predicate_type=self._predicate_type,
+                predicate=self._predicate,
+            )
+        except ValidationError as e:
+            raise ValueError(f"invalid statement: {e}")
+
+        return Statement(stmt.model_dump_json(by_alias=True).encode())
```

Comment:
A consideration here: technically we make no canonical/serialization guarantees other than "it's valid JSON", but we *could* do RFC 8785 for good measure here.